<a href="https://colab.research.google.com/github/juanpamoji/Repo1/blob/main/EntrenaTPU.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Entendiento Google Colab**
# **Juan pablo Mojica Macías**
# Diplomado en IA y AP


# **Ejemplo para TPU**

In [ ]:
import numpy as np
import os 
import gzip 

def datos (ruta, tipo="train"):
  ruta_categorias=os.path.join(ruta, '%s-labels-idx1-ubyte.gz' % tipo)
  ruta_imagenes=os.path.join(ruta, '%s-images-idx3-ubyte.gz'% tipo)

  with gzip.open(ruta_categorias,'rb') as rut_cat:
    etiquetas=np.frombuffer(rut_cat.read(), dtype=np.uint8, offset=8)

  with gzip.open(ruta_imagenes,'rb') as rut_imgs:
    imagenes=np.frombuffer(rut_imgs.read(),dtype=np.uint8, offset=16).reshape(len(etiquetas),784)

  return imagenes, etiquetas

# **Lectura de datos en Drive**


In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

ruta='gdrive/My Drive/Trabajos Diplomado IA/fashion_mnist_data'

X_train, Y_train=datos(ruta,tipo='train')
X_test, Y_test=datos(ruta,tipo='test')

Mounted at /content/gdrive


# **Reajuste del tamaño de datos (Multiplo de 128)**

In [ ]:
n_train=len(X_train)
n_test=len(X_train)
mod_train=60000-(60000%128)
mod_test=10000-(10000%128)

X_train=X_train[0:mod_train,:]
X_test=X_test[0:mod_test,:]
Y_train=Y_train[0:mod_train]
Y_test=Y_test[0:mod_test]

X_train=np.reshape(X_train,(mod_train,28,28,1))
X_test=np.reshape(X_test,(mod_test,28,28,1))

# Importación de ***Tensorflow 2*** plus ***Keras***

In [ ]:
%tensorflow_version 2.x
import tensorflow as tf

print('Versión de Tensorflow: ' + tf.__version__)

Versión de Tensorflow: 2.4.1


# **Configuración de TPU**

In [ ]:
try:
  tpu = tf.distribute.cluster_resolver.TPUClusterResolver(tpu='grpc://' + os.environ['COLAB_TPU_ADDR'])  # Detectar TPU
  print('TPU encontrada ', tpu.cluster_spec().as_dict()['worker'])
except ValueError:
  raise BaseException('ERROR: TPU no encontrada!')

tf.config.experimental_connect_to_cluster(tpu)
tf.tpu.experimental.initialize_tpu_system(tpu)
tpu_strategy = tf.distribute.experimental.TPUStrategy(tpu)

# **Implementación del Modelo usando "tpu_strategy"**

In [ ]:
tf.random.set_seed(200)

with tpu_strategy.scope():
  model = tf.keras.models.Sequential() #Contenedor del modelo

  model.add(tf.keras.layers.BatchNormalization(input_shape=X_train.shape[1:])) #Normalización primera Capa
  model.add(tf.keras.layers.Conv2D(64, (5, 5), padding='same', activation='elu')) #Filtro convolucional
  model.add(tf.keras.layers.MaxPooling2D(pool_size=(2, 2), strides=(2,2))) #Capa maxpooling
  model.add(tf.keras.layers.Dropout(0.25)) #reduce el overfitting del modelo
  
  model.add(tf.keras.layers.BatchNormalization(input_shape=X_train.shape[1:])) #Normalización segunda  Capa
  model.add(tf.keras.layers.Conv2D(128, (5, 5), padding='same', activation='elu')) #Filtro convolucional
  model.add(tf.keras.layers.MaxPooling2D(pool_size=(2, 2))) #Capa maxpooling
  model.add(tf.keras.layers.Dropout(0.25)) #reduce el overfitting del modelo
  
  model.add(tf.keras.layers.BatchNormalization(input_shape=X_train.shape[1:]))#Normalización tercera Capa
  model.add(tf.keras.layers.Conv2D(256, (5, 5), padding='same', activation='elu')) #Filtro convolucional
  model.add(tf.keras.layers.MaxPooling2D(pool_size=(2, 2), strides=(2,2)))#Capa maxpooling
  model.add(tf.keras.layers.Dropout(0.25)) #reduce el overfitting del modelo

  model.add(tf.keras.layers.Flatten()) #Aplana datos
  model.add(tf.keras.layers.Dense(256)) #Agragamos una red neurona con 256 neuronas
  model.add(tf.keras.layers.Activation('elu'))
  model.add(tf.keras.layers.Dropout(0.5))
  model.add(tf.keras.layers.Dense(10)) #conexión de salida con 10 neuronas
  model.add(tf.keras.layers.Activation('softmax'))
  
  model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

  model.summary()

# **Entrenamiento con TPU**

In [ ]:
# Entrenamiento GPU
import timeit

x_train = X_train.astype(np.float32)
y_train = Y_train.astype(np.float32)
x_test = X_test.astype(np.float32)
y_test = Y_test.astype(np.float32)

def tpu_train():
  model.fit(x_train,y_train,validation_data=(x_test,y_test),batch_size=128,epochs=2,verbose=1)
  
  return None

tpu_time = timeit.timeit('tpu_train()', number=1, setup='from __main__ import tpu_train')

In [ ]:
print('Tiempo de entrenamiento: ' + str(tpu_time) + ' segundos')

Tiempo de entrenamiento: 24.961727596999935 segundos


# **Nota:**

El entrenamiento con **TPU** se demoró aproximadamente 24.96 segundos 